docs/architecture/adr-061-mempool-refactor.md

```diff
@@ -0,0 +1,89 @@
+# ADR 061: Generalize Mempool Implementation
+
+## Changelog
+
+- 2020-10-01: Initial version. (@ninjaahhh)
+
+## Context
+
+When designing a [priority-based mempool](https://github.com/tendermint/spec/pull/154), we realized the current mempool interface needs to be refactored to support different underlying data structures to be priority-aware (e.g. max-heap, balance tree, etc.). This document lists necessary changes on the instance (i.e. struct)definition.
+
+## Decision
+
+[TBD]
+
+## Detailed Design
+
+Existing mempool code has many places tightly coupled with `CList` implementation (like in mempool reactor), and this proposed change mainly works as a middle ground to abstract away this coupling to allow future mempool implementation using other data structures for better performances related to transaction priorities.
+
+```golang
+type basemempool struct {
+    // Unified interface for actual mempool implementations
+    mempoolImpl
+
+    // Atomic integers
+    height   int64 // the last block Update()'d to
+    txsBytes int64 // total size of mempool, in bytes
+
+    // Notify listeners (ie. consensus) when txs are available
+    notifiedTxsAvailable bool
+    txsAvailable         chan struct{} // fires once for each height, when the mempool is not empty
+
+    // Keep a cache of already-seen txs.
+    // This reduces the pressure on the proxyApp.
+    cache     txCache
+    preCheck  PreCheckFunc
+    postCheck PostCheckFunc
+
+    config *cfg.MempoolConfig
+
+    // Exclusive mutex for Update method to prevent concurrent execution of
+    // CheckTx or ReapMaxBytesMaxGas(ReapMaxTxs) methods.
+    updateMtx sync.RWMutex
+
+    wal          *auto.AutoFile // a log of mempool txs
+    proxyAppConn proxy.AppConnMempool
+
+    logger  log.Logger
+    metrics *Metrics
+}
+
+type mempoolImpl interface {
```

Comment:
maybe we can find a better name for this, like `mempoolList`, which better reflects the underlying structure and its purpose.